pkg/plugincmdtree/plugins_cache.go

```diff
@@ -0,0 +1,257 @@
+// Copyright 2023 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package plugincmdtree
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
+	"gopkg.in/yaml.v2"
+
+	"github.com/vmware-tanzu/tanzu-cli/pkg/cli"
+	"github.com/vmware-tanzu/tanzu-cli/pkg/common"
+)
+
+const pluginsCommandTreeDir = "plugins_command_tree"
+
+type pluginKey struct {
+	PluginName string `yaml:"pluginName" json:"pluginName"`
+	Target     string `yaml:"target" json:"target"`
+	Version    string `yaml:"version" json:"version"`
+}
+
+type pluginCommandTree struct {
+	CommandTree map[pluginKey]*CommandNode `yaml:"commandTree" json:"commandTree"`
+}
+
+func getPluginsCommandTreeCacheDir() string {
+	// NOTE: TEST_CUSTOM_PLUGIN_COMMAND_TREE_CACHE_DIR is only for test purpose
+	customCommandTreeCacheDirForTest := os.Getenv("TEST_CUSTOM_PLUGIN_COMMAND_TREE_CACHE_DIR")
+	if customCommandTreeCacheDirForTest != "" {
+		return customCommandTreeCacheDirForTest
+	}
+	return filepath.Join(common.DefaultCacheDir, pluginsCommandTreeDir)
+}
+func GetPluginsCommandTreeCachePath() string {
+	return filepath.Join(getPluginsCommandTreeCacheDir(), "command_tree.yaml")
+}
+
+func getPluginsDocsCachePath() string {
+	return filepath.Join(getPluginsCommandTreeCacheDir(), "docs")
+}
+
+type cacheImpl struct {
+	pluginCommands      *pluginCommandTree
+	pluginDocsGenerator func(plugin *cli.PluginInfo) error
+}
+
+func NewCache() (Cache, error) {
+	pct, err := getPluginCommandTree()
+	if err != nil {
+		return nil, err
+	}
+	return &cacheImpl{
+		pluginCommands:      pct,
+		pluginDocsGenerator: generatePluginDocs,
```

Comment:
Nit: somewhere in the cacheImpl definition or implementation, we should briefly explain the approach taken here of leveraging generate-docs and parsing its output. The parsing adds complexity, and might get even messier when we need to learn more about each command beyond its subcommands and aliases. Another alternative would be to introduce plugin-runtime changes generate an even more structure and detailed output of a plugin's command tree.

Reply:
Sure. Added the brief explanation on the approach taken and the possible future enhancement option to teach plugin through plugin-runtime library to return command tree with exhaustive details that can be used for parsing.